quickwit-indexing/src/source/kinesis/api.rs

```diff
@@ -0,0 +1,449 @@
+    P: Fn(&str) -> bool,
+{
+    tokio::time::timeout(timeout, async {
+        let period = Duration::from_millis(if cfg!(test) { 100 } else { 5000 });
+        let mut interval = tokio::time::interval(period);
+        loop {
+            interval.tick().await;
+            let stream_status = describe_stream(kinesis_client, stream_name)
+                .await?
+                .stream_status;
+            if stream_status_predicate(&stream_status) {
+                return Ok(());
+            }
+        }
+    })
+    .await
+}
+
+#[cfg(all(test, feature = "kinesis-external-service"))]
+mod kinesis_localstack_tests {
+    use std::collections::HashMap;
+
+    use quickwit_common::rand::append_random_suffix;
+    use rusoto_core::Region;
+    use rusoto_kinesis::{KinesisClient, PutRecordsInput, PutRecordsRequestEntry};
+
+    use super::*;
+
+    fn get_localstack_client() -> KinesisClient {
+        KinesisClient::new(Region::Custom {
+            name: "localstack".to_string(),
+            endpoint: "http://localhost:4566".to_string(),
+        })
+    }
+
+    fn make_shard_id(id: usize) -> String {
+        format!("shardId-{:0>12}", id)
+    }
+
+    fn parse_shard_id<S: AsRef<str>>(shard_id: S) -> Option<usize> {
+        shard_id
+            .as_ref()
+            .strip_prefix("shardId-")
+            .and_then(|shard_id| shard_id.parse::<usize>().ok())
+    }
+
+    async fn put_records_into_shards<I>(
+        kinesis_client: &dyn Kinesis,
+        stream_name: &str,
+        records: I,
+    ) -> anyhow::Result<HashMap<usize, Vec<String>>>
+    where
+        I: IntoIterator<Item = (usize, &'static str)>,
+    {
+        let shard_hash_keys: HashMap<usize, String> =
+            list_shards(kinesis_client, &stream_name, None)
+                .await?
+                .into_iter()
+                .map(|shard| {
+                    (
+                        parse_shard_id(shard.shard_id).unwrap(),
+                        shard.hash_key_range.starting_hash_key,
+                    )
+                })
+                .collect();
+
+        let put_records_request_entries = records
+            .into_iter()
+            .map(|(shard_id, record)| PutRecordsRequestEntry {
+                explicit_hash_key: shard_hash_keys.get(&shard_id).cloned(),
+                partition_key: "Overridden by hash key".to_string(),
+                data: bytes::Bytes::from(record),
+            })
+            .collect();
+
+        let request = PutRecordsInput {
+            stream_name: stream_name.to_string(),
+            records: put_records_request_entries,
+        };
+        let response = kinesis_client.put_records(request).await?;
+
+        let mut sequence_numbers = HashMap::new();
+        for record in response.records {
+            sequence_numbers
+                .entry(record.shard_id.and_then(parse_shard_id).unwrap())
+                .or_insert_with(Vec::new)
+                .push(record.sequence_number.unwrap());
```

Comment:
I don't think we should ever unwrap things that are coming from a third-party.
In an ideal world, rusoto should take care of returning an error if the field cannot be deserialized but they don't because it is generated from metadata that does not contain this information.

In this non ideal world, unwrapping here is as dangerous as unwrapping during deserialization in the rusoto library.

Reply:
I agree with you. I'm unwrapping here because this function is only used for unit testing.